File: providers/zoominfo.py
"""ZoomInfo provider for Enterprise contact enrichment."""

import requests
from config import ZOOMINFO_CLIENT_ID, ZOOMINFO_CLIENT_SECRET

ZOOMINFO_AUTH_URL = "https://api.zoominfo.com/authenticate"
ZOOMINFO_ENRICH_URL = "https://api.zoominfo.com/enrich/company"
ZOOMINFO_CONTACTS_URL = "https://api.zoominfo.com/search/contact"

_access_token = None


def _get_access_token() -> str:
    """Get or refresh ZoomInfo access token."""
    global _access_token

    if not ZOOMINFO_CLIENT_ID or not ZOOMINFO_CLIENT_SECRET:
        raise ValueError("ZOOMINFO_CLIENT_ID and ZOOMINFO_CLIENT_SECRET not configured")

    if _access_token:
        return _access_token

    response = requests.post(
        ZOOMINFO_AUTH_URL,
        json={
            "username": ZOOMINFO_CLIENT_ID,
            "password": ZOOMINFO_CLIENT_SECRET,
        },
        timeout=30,
    )
    response.raise_for_status()

    data = response.json()
    _access_token = data.get("jwt")
    return _access_token


def enrich_company(domain: str) -> dict | None:
    """
    Enrich a company and get contacts using ZoomInfo.

    Args:
        domain: Company domain (e.g., "salesforce.com")

    Returns:
        Dict with company info and contacts, or None if not found
    """
    token = _get_access_token()
    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json",
    }

    # Enrich company
    company_response = requests.post(
        ZOOMINFO_ENRICH_URL,
        headers=headers,
        json={
            "matchCompanyInput": [{"companyWebsite": domain}],
            "outputFields": [
                "id", "name", "website", "industry", "employeeCount",
                "revenue", "city", "state", "country", "description"
            ],
        },
        timeout=30,
    )
    company_response.raise_for_status()

    company_data = company_response.json()
    companies = company_data.get("data", {}).get("result", [])

    if not companies:
        return None

    company = companies[0].get("data", [{}])[0] if companies[0].get("data") else {}

    # Search for contacts at this company
    contacts_response = requests.post(
        ZOOMINFO_CONTACTS_URL,
        headers=headers,
        json={
            "companyWebsite": [domain],
            "jobFunction": ["C-Level", "VP-Level", "Director"],
            "rpp": 10,
            "outputFields": [
                "id", "firstName", "lastName", "jobTitle", "email",
                "phone", "directPhoneNumber", "linkedinUrl"
            ],
        },
        timeout=30,
    )
    contacts_response.raise_for_status()

    contacts_data = contacts_response.json()
    contact_results = contacts_data.get("data", {}).get("result", [])

    contacts = []
    for contact in contact_results:
        contacts.append({
            "name": f"{contact.get('firstName', '')} {contact.get('lastName', '')}".strip(),
            "title": contact.get("jobTitle", ""),
            "email": contact.get("email", ""),
            "phone": contact.get("phone", ""),
            "direct_dial": contact.get("directPhoneNumber", ""),
            "linkedin_url": contact.get("linkedinUrl", ""),
        })

    return {
        "company": {
            "name": company.get("name", ""),
            "domain": company.get("website", domain),
            "industry": company.get("industry", ""),
            "employee_count": company.get("employeeCount"),
            "revenue": _format_revenue(company.get("revenue")),
            "city": company.get("city", ""),
            "state": company.get("state", ""),
            "country": company.get("country", ""),
            "description": company.get("description", ""),
        },
        "contacts": contacts,
    }


def _format_revenue(revenue) -> str:
    """Format revenue as readable string."""
    if not revenue:
        return "Unknown"
    try:
        revenue = float(revenue)
        if revenue >= 1_000_000_000:
            return f"${revenue / 1_000_000_000:.1f}B"
        if revenue >= 1_000_000:
            return f"${revenue / 1_000_000:.0f}M"
        if revenue >= 1_000:
            return f"${revenue / 1_000:.0f}K"
        return f"${revenue:.0f}"
    except (ValueError, TypeError):
        return str(revenue)

File: frontend/lib/providers/zoominfo.ts
/**
 * ZoomInfo Provider Adapter
 *
 * Enterprise B2B intelligence platform with JWT authentication.
 * Provides comprehensive company and contact data.
 *
 * API Docs: https://www.zoominfo.com/solutions/integrations
 */

import {
  ProviderAdapter,
  ProviderResponse,
  ContactResponse,
  CompanyQuery,
  ContactQuery,
  ProviderError,
  createProviderResponse,
  createContactResponse,
  formatRevenue,
} from './types';

const ZOOMINFO_AUTH_URL = 'https://api.zoominfo.com/authenticate';
const ZOOMINFO_ENRICH_URL = 'https://api.zoominfo.com/enrich/company';
const ZOOMINFO_CONTACTS_URL = 'https://api.zoominfo.com/search/contact';

// Module-level token cache (simple PoC implementation)
let cachedAccessToken: string | null = null;

/**
 * Get ZoomInfo credentials from environment.
 */
function getCredentials(): { clientId: string; clientSecret: string } {
  const clientId = process.env.ZOOMINFO_CLIENT_ID;
  const clientSecret = process.env.ZOOMINFO_CLIENT_SECRET;

  if (!clientId || !clientSecret) {
    throw new ProviderError(
      'zoominfo',
      'config_error',
      'ZOOMINFO_CLIENT_ID and ZOOMINFO_CLIENT_SECRET not configured'
    );
  }

  return { clientId, clientSecret };
}

/**
 * Get or refresh ZoomInfo JWT access token.
 */
async function getAccessToken(): Promise<string> {
  if (cachedAccessToken) {
    return cachedAccessToken;
  }

  const { clientId, clientSecret } = getCredentials();

  const response = await fetch(ZOOMINFO_AUTH_URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      username: clientId,
      password: clientSecret,
    }),
  });

  if (!response.ok) {
    throw new ProviderError(
      'zoominfo',
      'auth_error',
      `Authentication failed: ${response.statusText}`,
      response.status
    );
  }

  const data = await response.json();
  cachedAccessToken = data.jwt;

  if (!cachedAccessToken) {
    throw new ProviderError('zoominfo', 'auth_error', 'No JWT token in response');
  }

  return cachedAccessToken;
}

/**
 * Clear cached access token (useful for testing or token refresh).
 */
export function clearTokenCache(): void {
  cachedAccessToken = null;
}

/**
 * ZoomInfo Provider Adapter
 *
 * Enterprise-grade company and contact enrichment with JWT authentication.
 */
export class ZoomInfoAdapter implements ProviderAdapter {
  id = 'zoominfo';
  name = 'ZoomInfo';

  /**
   * Enrich a company and optionally get executive contacts.
   */
  async enrichCompany(query: CompanyQuery): Promise<ProviderResponse | null> {
    if (!query.domain) {
      throw new ProviderError('zoominfo', 'api_error', 'Domain is required for ZoomInfo enrichment');
    }

    const token = await getAccessToken();
    const headers = {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    };

    // Enrich company
    const companyResponse = await fetch(ZOOMINFO_ENRICH_URL, {
      method: 'POST',
      headers,
      body: JSON.stringify({
        matchCompanyInput: [{ companyWebsite: query.domain }],
        outputFields: [
          'id',
          'name',
          'website',
          'industry',
          'employeeCount',
          'revenue',
          'city',
          'state',
          'country',
          'description',
        ],
      }),
    });

    if (!companyResponse.ok) {
      // Clear token cache on auth errors
      if (companyResponse.status === 401) {
        clearTokenCache();
      }
      throw new ProviderError(
        'zoominfo',
        'api_error',
        `Company enrichment failed: ${companyResponse.statusText}`,
        companyResponse.status
      );
    }

    const companyData = await companyResponse.json();
    const companies = companyData.data?.result || [];

    if (companies.length === 0) {
      return null;
    }

    const company = companies[0]?.data?.[0] || {};

    // Search for executive contacts at this company
    const contactsResponse = await fetch(ZOOMINFO_CONTACTS_URL, {
      method: 'POST',
      headers,
      body: JSON.stringify({
        companyWebsite: [query.domain],
        jobFunction: ['C-Level', 'VP-Level', 'Director'],
        rpp: 10,
        outputFields: [
          'id',
          'firstName',
          'lastName',
          'jobTitle',
          'email',
          'phone',
          'directPhoneNumber',
          'linkedinUrl',
        ],
      }),
    });

    if (!contactsResponse.ok) {
      throw new ProviderError(
        'zoominfo',
        'api_error',
        `Contact search failed: ${contactsResponse.statusText}`,
        contactsResponse.status
      );
    }

    const contactsData = await contactsResponse.json();
    const contactResults = contactsData.data?.result || [];

    const contacts = contactResults.map((contact: Record<string, unknown>) => ({
      name: `${contact.firstName || ''} ${contact.lastName || ''}`.trim(),
      title: contact.jobTitle || '',
      email: contact.email || '',
      phone: contact.phone || '',
      direct_dial: contact.directPhoneNumber || '',
      linkedin_url: contact.linkedinUrl || '',
    }));

    return createProviderResponse(
      this.id,
      { company, contacts: contactResults },
      {
        company: {
          name: company.name || '',
          domain: company.website || query.domain,
          industry: company.industry || '',
          employee_count: company.employeeCount,
          revenue: formatRevenue(company.revenue),
          city: company.city || '',
          state: company.state || '',
          country: company.country || '',
          description: company.description || '',
        },
        contacts,
      }
    );
  }

  /**
   * Search for contacts at a company.
   */
  async searchContacts(query: ContactQuery): Promise<ContactResponse[]> {
    const token = await getAccessToken();

    const payload: Record<string, unknown> = {
      companyWebsite: [query.domain],
      rpp: query.limit || 10,
      outputFields: [
        'id',
        'firstName',
        'lastName',
        'jobTitle',
        'email',
        'phone',
        'directPhoneNumber',
        'linkedinUrl',
      ],
    };

    // Map titles to ZoomInfo job functions if provided
    if (query.titles && query.titles.length > 0) {
      payload.jobTitle = query.titles;
    }

    const response = await fetch(ZOOMINFO_CONTACTS_URL, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      if (response.status === 401) {
        clearTokenCache();
      }
      throw new ProviderError(
        'zoominfo',
        'api_error',
        `Contact search failed: ${response.statusText}`,
        response.status
      );
    }

    const data = await response.json();
    const contacts = data.data?.result || [];

    return contacts.map((contact: Record<string, unknown>) =>
      createContactResponse(this.id, {
        name: `${contact.firstName || ''} ${contact.lastName || ''}`.trim(),
        title: (contact.jobTitle as string) || '',
        email: (contact.email as string) || '',
        phone: (contact.directPhoneNumber as string) || (contact.phone as string) || '',
        linkedin_url: (contact.linkedinUrl as string) || '',
        raw: contact,
      })
    );
  }
}

export default ZoomInfoAdapter;
